share/man/man4/puc.4
.\" $NetBSD$
.\"
.\" Copyright (c) 1998 Christopher G. Demetriou
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"          This product includes software developed for the
.\"          NetBSD Project.  See http://www.NetBSD.org/ for
.\"          information about NetBSD.
.\" 4. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" <<Id: LICENSE,v 1.2 2000/06/14 15:57:33 cgd Exp>>
.\"
.Dd May 4, 2015
.Dt PUC 4
.Os
.Sh NAME
.Nm puc
.Nd PCI
.Dq universal
communications card driver
.Sh SYNOPSIS
.Cd "puc* at pci? dev ? function ?"
.Cd "com* at puc? port ?"
.Cd "lpt* at puc? port ?"
.Sh DESCRIPTION
The
.Nm
driver provides support for PCI communications cards containing
simple communications ports, such as NS16550-family
.Pf ( Nm com )
serial ports and standard PC-like
.Pf ( Nm lpt )
parallel ports.
The driver is called
.Dq universal
because the interfaces to these devices aren't nearly as well
defined and standard as they should be.
.Pp
The driver currently supports the following cards:
.Pp
.Bl -tag -width Dv -offset indent -compact
.It Tn "ADDI-DATA APCI-7800 (8 port serial)"
.It Tn "Actiontec 56K PCI Master"
.It Tn "Advantech PCI-1604UP (2 port serial)"
.It Tn "Advantech PCI-1610 (4 port serial)"
.It Tn "Advantech PCI-1612 (4 port serial)"
.It Tn "Advantech PCI-1620 (8 port serial)"
.It Tn "Avlab Low Profile PCI 4S Quartet (4 port serial)"
.It Tn "Avlab Low Profile PCI 4 Serial (4 port serial)"
.It Tn "Avlab PCI 2S (2 port serial)"
.It Tn "Boca Research Turbo Serial 654 (4 port serial)"
.It Tn "Boca Research Turbo Serial 658 (8 port serial)"
.It Tn "Chase Research / Perle PCI-FAST4 (4 port serial)"
.It Tn "Chase Research / Perle PCI-FAST8 (8 port serial)"
.It Tn "Comtrol RocketPort 550/4 series (4 port serial)"
.It Tn "Comtrol RocketPort 550/8 series (8 port serial)"
.It Tn "Comtrol RocketPort 550/16 series (16 port serial)"
.It Tn "Decision Computer Inc PCCOM PCI 2 Port (2 port serial)"
.It Tn "Decision Computer Inc PCCOM PCI 4 Port (4 port serial)"
.It Tn "Decision Computer Inc PCCOM PCI 8 Port (8 port serial)"
.It Tn "Digi International Digi Neo 4 (4 port serial)"
.It Tn "Digi International Digi Neo 8 (8 port serial)"
.It Tn "Dolphin Peripherals 4014 (dual parallel)"
.It Tn "Dolphin Peripherals 4035 (dual serial)"
.It Tn "Dolphin Peripherals 4036 (dual serial)"
.It Tn "EXAR XR17D152 (2 port serial)"
.It Tn "EXAR XR17D154 (4 port serial)"
.It Tn "EXAR XR17D158 (8 port serial)"
.It Tn "Exsys EX-41098 (4 port serial)"
.It Tn "IBM 4810 SurePOS 300 Series SCC (4 port serial)"
.It Tn "InnoSys Keyspan SX Pro (4 port serial)"
.It Tn "Intel chipset internal Serial over LAN"
.It Tn "I-O DATA RSA-PCI (2 port serial)"
.It Tn "I-O DATA RSA-PCI2 (2 port serial)"
.It Tn "I-O DATA RSA-PCI2/P4 (4 port serial)"
.It Tn "I-O DATA RSA-PCI2/P8 (8 port serial)"
.It Tn "Lava Computers 2SP-PCI (single parallel)"
.It Tn "Lava Computers Octopus (8 port serial)"
.It Tn "Lava Computers Quatro-PCI (4 port serial)"
.It Tn "Lava Computers dual serial"
.It Tn "Lava Computers SSERIAL-PCI (single serial)"
.It Tn "Middle Digital, Inc. Weasel serial port"
.It Tn "Moxa Technologies SmartIO C104H/PCI (4 port serial)"
.It Tn "Moxa Technologies SmartIO C168EL-A/PCIe (8 port serial)"
.It Tn "Moxa Technologies SmartIO C168EL/PCIe (8 port serial)"
.It Tn "Moxa Technologies SmartIO C168H/PCI (8 port serial)"
.It Tn "Moxa Technologies SmartIO C168U/PCI (8 port serial)"
.It Tn "Moxa Technologies SmartIO CP-114/PCI (4 port serial)"
.It Tn "Moxa Technologies SmartIO CP-102/PCI (2 port serial)"
.It Tn "Moxa Technologies SmartIO CP-104-EL/PCIe (4 port serial)"
.It Tn "Moxa Technologies SmartIO CP-104-V2/PCI (4 port serial)"
.It Tn "Moxa Technologies SmartIO CP-104/PCI (4 port serial)"
.It Tn "Multi-Tech ISI5634PCI/4 (4 port serial)"
.It Tn "NEC PK-UG-X001 K56flex PCI Modem"
.It Tn "NEC PK-UG-X008"
.It Tn "NetMos 1P PCI (single parallel)"
.It Tn "NetMos 2S1P PCI 16C650 (dual serial and single parallel)"
.It Tn "NetMos 4S1P PCI NM9845 (4 port serial and single parallel)"
.It Tn "NetMos NM9805 1284 (single parallel)"
.It Tn "NetMos NM9815 Dual 1284 (dual parallel)"
.It Tn "NetMos NM9835 series (up to dual serial and single parallel)"
.It Tn "NetMos NM9845 series (up to 6 serial and 1 parallel)"
.It Tn "NetMos NM9865 series (up to 4 serial and 2 parallel)"
.It Tn "NetMos NM9901 PCIe (1 serial or 1 parallel)"
.It Tn "NetMos NM9904 PCIe (4 port serial)"
.It Tn "NetMos NM9922 PCIe (2 port serial)"
.It Tn "Oxford Semiconductor OX16PCI952 (dual serial and single parallel)"
.It Tn "Oxford Semiconductor OX16PCI954 (4 port serial)"
.It Tn "Oxford Semiconductor OX16PCI958 (8 port serial)"
.It Tn "Oxford Semiconductor OXPCIe952 (2 port serial, legacy mode)"
.It Tn "Oxford Semiconductor OXmPCI952 (2 port serial)"
.It Tn "Perle Systems PCI-RAS 4 modem ports"
.It Tn "Perle Systems PCI-RAS 8 modem ports"
.It Tn "Perle Systems PCI-RASV92 4 modem ports"
.It Tn "Perle Systems PCI-RASV92 8 modem ports"
.It Tn "SIIG Cyber 2P1S PCI series (dual parallel and single serial)"
.It Tn "SIIG Cyber 2S1P PCI series (dual serial and single parallel)"
.It Tn "SIIG Cyber 4 PCI 16550 (4 port serial)"
.It Tn "SIIG Cyber 4S PCI series (quad serial)"
.It Tn "SIIG Cyber I/O PCI series (single serial and single parallel)"
.It Tn "SIIG Cyber Parallel Dual PCI series (dual parallel)"
.It Tn "SIIG Cyber Parallel PCI series (single parallel)"
.It Tn "SIIG Cyber Serial Dual PCI series (dual serial)"
.It Tn "SIIG Cyber Serial PCI series (single serial)"
.It Tn "SIIG PS8000 PCI 8S series (8 port serial)"
.It Tn "SUNIX 400x (1 port parallel)"
.It Tn "SUNIX 401x (2 port parallel)"
.It Tn "SUNIX 402x (1 port serial)"
.It Tn "SUNIX 403x (2 port serial)"
.It Tn "SUNIX 405x (4 port serial)"
.It Tn "SUNIX 406x (8 port serial)"
.It Tn "SUNIX 407x (2 port serial and 1 port parallel)"
.It Tn "SUNIX 408x (2 port serial and 2 port parallel)"
.It Tn "SUNIX 409x (4 port serial and 2 port parallel)"
.It Tn "SUNIX 5008 (1 port parallel)"
.It Tn "SUNIX 5016 (8 port serial)"
.It Tn "SUNIX 5027 (1 port serial)"
.It Tn "SUNIX 5037 (2 port serial)"
.It Tn "SUNIX 5056 (4 port serial)"
.It Tn "SUNIX 5066 (8 port serial)"
.It Tn "SUNIX 5069 (1 port serial and 1 port parallel)"
.It Tn "SUNIX 5079 (2 port serial and 1 port parallel)"
.It Tn "SUNIX 5099 (4 port serial and 1 port parallel)"
.It Tn "Syba Tech Ltd. PCI-4S"
.It Tn "Syba Tech Ltd. PCI-4S2P-550-ECP"
.It Tn "SystemBase SB16C1050PCI (2 port serial)"
.It Tn "SystemBase SB16C1054PCI (4 port serial)"
.It Tn "SystemBase SB16C1058PCI (8 port serial)"
.It Tn "US Robotics (3Com) 3CP5609 PCI 16550 Modem"
.It Tn "VScom PCI-010HV2 (1 port parallel)"
.It Tn "VScom PCI-010L (1 port parallel)"
.It Tn "VScom PCI-011H (1 port parallel)"
.It Tn "VScom PCI-100H (1 port serial)"
.It Tn "VScom PCI-100L (1 port serial)"
.It Tn "VScom PCI-110L (1 port serial and 1 port parallel)"
.It Tn "VScom PCI-200 (dual serial)"
.It Tn "VScom PCI-200H (dual serial)"
.It Tn "VScom PCI-200HV2 (dual serial)"
.It Tn "VScom PCI-200L (dual serial)"
.It Tn "VScom PCI-200Li (dual serial)"
.It Tn "VScom PCI-210L (2 port serial and 1 port parallel)"
.It Tn "VScom PCI-400 (4 port serial)"
.It Tn "VScom PCI-400L (4 port serial)"
.It Tn "VScom PCI-800 (8 port serial)"
.It Tn "VScom PCI-800H (8 port serial)"
.It Tn "VScom PCI-800L (8 port serial)"
.El
.Pp
The driver does not support the cards:
.Pp
.Bl -tag -width Fl -offset indent -compact
.It Tn "Dolphin Peripherals 4006 (single parallel)"
.It Tn "Dolphin Peripherals 4025 (single serial)"
.It Tn "Dolphin Peripherals 4078 (dual serial and single parallel)"
.El
.Pp
but support for them (and for similar cards) should be trivial to add.
.Pp
The
.Ar port
locator is used to identify the port (starting from 0) on the
communications card that a subdevice is supposed to attach to.
Typically, the numbering of ports is explained in a card's
hardware documentation, and the port numbers used by the driver
are the same as (or one off from, e.g. the manual uses ports
numbered starting from 1) those described in the documentation.
.Sh SEE ALSO
.Xr com 4 ,
.Xr lpt 4 ,
.Xr pci 4
.Sh HISTORY
The
.Nm
driver appeared in
.Nx 1.4 .
.Sh AUTHORS
The
.Nm
driver was written by
.An Chris Demetriou .
.Sh BUGS
The current design of this driver keeps any
.Nm com
ports on these cards from easily being used as console.
Of course, because boards with those are PCI boards, they also
suffer from dynamic address assignment, which also means that they
can't easily be used as console.
.Pp
Some of the cards supported by this driver have jumper-selectable
.Nm com
port clock multipliers, which are unsupported by this
driver.
Those can be easily accommodated with driver flags, or by
using a properly scaled baud rate when talking to the card.
.Pp
Some of the cards supported by this driver, e.g. the VScom PCI-800,
have software-selectable
.Nm com
port clock multipliers, which are unsupported by this driver.
Those can be accommodated using internal driver flags, or by using
a properly scaled baud rate when talking to the card.
.Pp
Some ports use an
.Nm lpt
driver other than the machine-independent driver.
Those ports will not be able to use
.Nm lpt
ports attached to
.Nm
devices.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 7.0 to NetBSD 8.0:
	libc: Update to tzcode2014f. [christos 20140815]
	gmake:	Update to gmake-3.81 in external/gpl2/gmake and remove 
		gnu/dist/gmake (3.80) [christos 20140818]
	module(7): Create MODULAR_DEFAULT_AUTOLOAD config flag to control
		whether module autoloading is on by default.
		[jnemeth 20140824]
	kernel: Option SYMTAB_SPACE has been replaced with COPY_SYMTAB.
		The related changes to src/sys/conf require a rebuild of
		config(1). [joerg 20140824]
	zoneinfo: Import tzdata2014g. [apb 20140831]
	wm(4): Add internal SERDES support for 82575 and newer controllers.
		[msaitoh 20140903]
	or1k: Add new port.  [matt 20140903]
	awindma(4): Add support for Allwinner A10/A20 DMA controller.
		[jmcneill 20140906]
	awinac(4): Add support for Allwinner A10/A20 audio codec.
		[jmcneill 20140906]
	awinrtc(4): Add support for Allwinner A10/A20 real-time clock.
		[jmcneill 20140907]
	awinmmc(4): Use DMA for MMC transfers. [jmcneill 20140908]
	axp20x(4): Add AXP20x PMU driver. [jmcneill 20140909]
	rtsol(8): Removed in favour of dhcpcd. [roy 20140911]
	rtsold(8): Removed in favour of dhcpcd. [roy 20140911]
	dhcpcd(8): Import dhcpcd-6.4.5. [roy 20140918]
	arm: Add support for i.MX6 SoC. [ryo 20140925]
	gpt(8): Completed overhaul, including adding follwing subcommands:
		resize, set, unset, backup, restore, and resizedisk.
		[jnemeth 20140926]
	dhcpcd(8): Import dhcpcd-6.4.7. [roy 20140927]
	mpt(4): added bio(4) support. [jmcneill 20140927]
	gpt(8): added type subcommand to modify partition type.
		[jnemeth 20140927]
	zoneinfo: Import tzdata2014h. [apb 20141004]
	dhcpcd(8): Import dhcpcd-6.5.0. [roy 20141006]
	libc: Update to tzcode2014h. [christos 20141007]
	file(1): Upgraded to 5.20. [christos 20141010]
	arm: Add support for Allwinner A31 SoC. [jmcneill 20141010]
	awiniic(4): Add support for Allwinner A31 SoC. [jmcneill 20141012]
	awinp2wi(4): Add support for Allwinner A31 Push-Pull Two Wire
		Interface. [jmcneill 20141012]
	axp22x(4): Add AXP22x PMU driver. [jmcneill 20141012]
	awinrtc(4): Add support for Allwinner A31 real-time clock.
		[jmcneill 20141012]
	awindma(4): Add support for Allwinner A31 DMA controller.
		[jmcneill 20141013]
	awinac(4): Add support for Allwinner A31 audio codec.
		[jmcneill 20141013]
	motg(4): Add support for Allwinner A10/A20 SoC. [jmcneill 20141015]
	wpa: Import wpa_supplicant and hostapd 2.3. [christos 20141016]
	dhcpcd(8): Import dhcpcd-6.5.1. [roy 20141018]
	OpenSSH: Imported 6.7. [christos 20141018]
	resolvconf(8): Import openresolv-3.6.0 [roy 20141020]
	gcc: Support -fsanitize=address [christos 20141022]
	zoneinfo: Import tzdata2014i. [apb 20141023]
	libc: Update to tzcode2014i. [christos 20141023]
	openpam(3): update to 20140912 (ourouparia) [christos 20141024]
	pppd(8): updated to version 2.4.7. [christos 20141025]
	acpi(4): Updated ACPICA to 20140926. [christos 20141025]
	dhcpcd(8): Import dhcpcd-6.6.0 [roy 20141029]
	resolvconf(8): Import openresolv-3.6.1 [roy 20141029]
	rtsx(4): Add support for Realtek RTS5227/RTL8402/RTL8411/RTL8411B.
		[nonaka 20141029]
	flex(1): Import flex-2.5.37 [christos 20141029]
	iwn(4): Add support for Intel Centrino Wireless-N
		100/105/130/135/2200/2230. From OpenBSD. [nonaka 20141030]
	awge(4): Add support for gigabit ethernet found on various
		Allwinner SOCs. [martin 20141101]
	dhcpcd(8): Import dhcpcd-6.6.1. [roy 20141107]
	awinhdmi(4): Add support for Allwinner A20/A31 HDMI controller.
		[jmcneill 20141110]
	awintcon(4): Add support for Allwinner A20/A31 TV/LCD controller.
		[jmcneill 20141110]
	awindebe(4): Add support for Allwinner A20/A31 Display engine
		backend (DE-BE). [jmcneill 20141110]
	awinhdmiaudio(4): Add support for Allwinner A20/A31 HDMI audio
		controller. [jmcneill 20141111]
	dhcpcd(8): Import dhcpcd-6.6.2. [roy 20141114]
	awinir(4): Add support for Allwinner A20/A31 IR receiver.
		[jmcneill 20141115]
	zoneinfo: Import tzdata2014j. [apb 20141116]
	libc: Change arc4random(3) to use ChaCha20 and per-thread state.
		[riastradh 20141116]
	libc: Update to tzcode2014j. [christos 20141117]
	libpcap: Import 1.6.2. [christos 20141119]
	tcpdump(8): Import 4.6.2. [christos 20141119]
	dhcpcd(8): Import dhcpcd-6.6.4. [roy 20141126]
	dhcpcd(8): Import dhcpcd-6.6.5. [roy 20141209]
	bind: Import version 9.10.1-P1. [christos 20141209]
	dhcpcd(8): Import dhcpcd-6.6.6. [roy 20141217]
	powerpc: Switch to GCC 4.8. [matt 20141218]
	dhcpcd(8): Import dhcpcd-6.6.7. [roy 20141219]
	ntp: Import ntp 4.2.8. [christos 20141219]
	midirecord(1): Add program to record MIDI files.  [mrg 20141229]
	file: update to 5.22 security limits [christos 20150102]
	byacc: update to 20141128 for more bison compat [christos 20150103]
	mgx: driver for Southland Media Systems (now Quantum 3D) MGX SBus
		graphics devices [macallan 20150106]
	gcc: Import GCC 4.8.4.  [mrg 20150109]
	openssl: Import openssl 1.0.1k [spz 20150113]
	am-utils: Upgrade to 6.2; adds nfsv4, lustre (which we can't use)
		nfs3 rpc support for amfs [christos 20150117]

sys/dev/pci/puc.c
/*	$NetBSD$	*/

/*
 * Copyright (c) 1996, 1998, 1999
 *	Christopher G. Demetriou.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by Christopher G. Demetriou
 *	for the NetBSD Project.
 * 4. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * PCI "universal" communication card device driver, glues com, lpt,
 * and similar ports to PCI via bridge chip often much larger than
 * the devices being glued.
 *
 * Author: Christopher G. Demetriou, May 14, 1998 (derived from NetBSD
 * sys/dev/pci/pciide.c, revision 1.6).
 *
 * These devices could be (and some times are) described as
 * communications/{serial,parallel}, etc. devices with known
 * programming interfaces, but those programming interfaces (in
 * particular the BAR assignments for devices, etc.) in fact are not
 * particularly well defined.
 *
 * After I/we have seen more of these devices, it may be possible
 * to generalize some of these bits.  In particular, devices which
 * describe themselves as communications/serial/16[45]50, and
 * communications/parallel/??? might be attached via direct
 * 'com' and 'lpt' attachments to pci.
 */

#include <sys/cdefs.h>
__KERNEL_RCSID(0, "$NetBSD$");

#include <sys/param.h>
#include <sys/systm.h>
#include <sys/device.h>

#include <dev/pci/pcireg.h>
#include <dev/pci/pcivar.h>
#include <dev/pci/pucvar.h>
#include <dev/pci/pcidevs.h>
#include <sys/termios.h>
#include <dev/ic/comreg.h>
#include <dev/ic/comvar.h>

#include "locators.h"
#include "com.h"

struct puc_softc {
	/* static configuration data */
	const struct puc_device_description *sc_desc;

	/* card-global dynamic data */
	void			*sc_ih;
	struct {
		int		mapped;
		bus_addr_t	a;
		bus_size_t	s;
		bus_space_tag_t	t;
		bus_space_handle_t h;
	} sc_bar_mappings[6];				/* XXX constant */

	/* per-port dynamic data */
        struct {
		device_t 	dev;

                /* filled in by port attachments */
                int             (*ihand)(void *);
                void            *ihandarg;
        } sc_ports[PUC_MAX_PORTS];
};

static int	puc_print(void *, const char *);

static const char *puc_port_type_name(int);

static int
puc_match(device_t parent, cfdata_t match, void *aux)
{
	struct pci_attach_args *pa = aux;
	const struct puc_device_description *desc;
	pcireg_t bhlc, subsys;

	bhlc = pci_conf_read(pa->pa_pc, pa->pa_tag, PCI_BHLC_REG);
	if (PCI_HDRTYPE_TYPE(bhlc) != 0)
		return (0);

	subsys = pci_conf_read(pa->pa_pc, pa->pa_tag, PCI_SUBSYS_ID_REG);

	desc = puc_find_description(PCI_VENDOR(pa->pa_id),
	    PCI_PRODUCT(pa->pa_id), PCI_VENDOR(subsys), PCI_PRODUCT(subsys));
	if (desc != NULL)
		return (10);

#if 0
	/*
	 * XXX this is obviously bogus.  eventually, we might want
	 * XXX to match communications/modem, etc., but that needs some
	 * XXX special work in the match fn.
	 */
	/*
	 * Match class/subclass, so we can tell people to compile kernel
	 * with options that cause this driver to spew.
	 */
	if (PCI_CLASS(pa->pa_class) == PCI_CLASS_COMMUNICATIONS) {
		switch (PCI_SUBCLASS(pa->pa_class)) {
		case PCI_SUBCLASS_COMMUNICATIONS_SERIAL:
		case PCI_SUBCLASS_COMMUNICATIONS_MODEM:
			return (1);
		}
	}
#endif

	return (0);
}

static void
puc_attach(device_t parent, device_t self, void *aux)
{
	struct puc_softc *sc = device_private(self);
	struct pci_attach_args *pa = aux;
	struct puc_attach_args paa;
	pci_intr_handle_t intrhandle;
	pcireg_t subsys;
	int i, barindex;
	int locs[PUCCF_NLOCS];

	subsys = pci_conf_read(pa->pa_pc, pa->pa_tag, PCI_SUBSYS_ID_REG);
	sc->sc_desc = puc_find_description(PCI_VENDOR(pa->pa_id),
	    PCI_PRODUCT(pa->pa_id), PCI_VENDOR(subsys), PCI_PRODUCT(subsys));
	if (sc->sc_desc == NULL) {
		/*
		 * This was a class/subclass match, so tell people to compile
		 * kernel with options that cause this driver to spew.
		 */
#ifdef PUC_PRINT_REGS
		printf(":\n");
		pci_conf_print(pa->pa_pc, pa->pa_tag, NULL);
#else
		printf(": unknown PCI communications device\n");
		printf("%s: compile kernel with PUC_PRINT_REGS and larger\n",
		    device_xname(self));
		printf("%s: message buffer (via 'options MSGBUFSIZE=...'),\n",
		    device_xname(self));
		printf("%s: and report the result with send-pr\n",
		    device_xname(self));
#endif
		return;
	}

	aprint_naive("\n");
	aprint_normal(": %s (", sc->sc_desc->name);
	for (i = 0; PUC_PORT_VALID(sc->sc_desc, i); i++)
		aprint_normal("%s%s", i ? ", " : "",
		    puc_port_type_name(sc->sc_desc->ports[i].type));
	aprint_normal(")\n");

	for (i = 0; i < 6; i++) {
		pcireg_t bar, type;

		sc->sc_bar_mappings[i].mapped = 0;

		bar = pci_conf_read(pa->pa_pc, pa->pa_tag, PCI_BAR(i));
		if (bar == 0)			/* BAR not implemented(?) */
			continue;

		type = (PCI_MAPREG_TYPE(bar) == PCI_MAPREG_TYPE_IO ?
		    PCI_MAPREG_TYPE_IO : PCI_MAPREG_MEM_TYPE(bar));

		if (type == PCI_MAPREG_TYPE_IO) {
			sc->sc_bar_mappings[i].t = pa->pa_iot;
			sc->sc_bar_mappings[i].a = PCI_MAPREG_IO_ADDR(bar);
			sc->sc_bar_mappings[i].s = PCI_MAPREG_IO_SIZE(bar);
		} else {
			sc->sc_bar_mappings[i].t = pa->pa_memt;
			sc->sc_bar_mappings[i].a = PCI_MAPREG_MEM_ADDR(bar);
			sc->sc_bar_mappings[i].s = PCI_MAPREG_MEM_SIZE(bar);
		}

		sc->sc_bar_mappings[i].mapped = (pci_mapreg_map(pa,
		    PCI_BAR(i), type, 0,
		    &sc->sc_bar_mappings[i].t, &sc->sc_bar_mappings[i].h,
		    &sc->sc_bar_mappings[i].a, &sc->sc_bar_mappings[i].s)
		      == 0);
		if (sc->sc_bar_mappings[i].mapped)
			continue;

		aprint_debug_dev(self, "couldn't map BAR at offset 0x%lx\n",
		    (long)(PCI_MAPREG_START + 4 * i));
	}

	/* Map interrupt. */
	if (pci_intr_map(pa, &intrhandle)) {
		aprint_error_dev(self, "couldn't map interrupt\n");
		return;
	}
	/*
	 * XXX the sub-devices establish the interrupts, for the
	 * XXX following reasons:
	 * XXX
	 * XXX    * we can't really know what IPLs they'd want
	 * XXX
	 * XXX    * the MD dispatching code can ("should") dispatch
	 * XXX      chained interrupts better than we can.
	 * XXX
	 * XXX It would be nice if we could indicate to the MD interrupt
	 * XXX handling code that the interrupt line used by the device
	 * XXX was a PCI (level triggered) interrupt.
	 * XXX
	 * XXX It's not pretty, but hey, what is?
	 */

	/* SB16C10xx board specific initialization */
	if (PCI_VENDOR(pa->pa_id) == PCI_VENDOR_SYSTEMBASE &&
	    (PCI_PRODUCT(pa->pa_id) == PCI_PRODUCT_SYSTEMBASE_SB16C1050 ||
	    PCI_PRODUCT(pa->pa_id) == PCI_PRODUCT_SYSTEMBASE_SB16C1054 ||
	    PCI_PRODUCT(pa->pa_id) == PCI_PRODUCT_SYSTEMBASE_SB16C1058)) {
		if (!sc->sc_bar_mappings[1].mapped) {
			aprint_error_dev(self,
			    "optional register is not mapped\n");
			return;
		}
#define SB16C105X_OPT_IMRREG0 0x0000000c
		/* enable port 0-7 interrupt */
		bus_space_write_1(sc->sc_bar_mappings[1].t,
		    sc->sc_bar_mappings[1].h, SB16C105X_OPT_IMRREG0, 0xff);
	} else {
		if (!pmf_device_register(self, NULL, NULL))
	                aprint_error_dev(self,
			    "couldn't establish power handler\n");
	}

	/* Configure each port. */
	for (i = 0; PUC_PORT_VALID(sc->sc_desc, i); i++) {
		barindex = PUC_PORT_BAR_INDEX(sc->sc_desc->ports[i].bar);
		bus_space_handle_t subregion_handle;
		int is_console = 0;

		/* make sure the base address register is mapped */
#if NCOM > 0
		is_console = com_is_console(sc->sc_bar_mappings[barindex].t,
		    sc->sc_bar_mappings[barindex].a +
		    sc->sc_desc->ports[i].offset, &subregion_handle);
		if (is_console) {
			sc->sc_bar_mappings[barindex].mapped = 1;
#if defined(amd64) || defined(i386)
			sc->sc_bar_mappings[barindex].h = subregion_handle -
			    sc->sc_desc->ports[i].offset;	/* XXX hack */
#endif
		}
#endif
		if (!sc->sc_bar_mappings[barindex].mapped) {
			aprint_error_dev(self,
			    "%s port uses unmapped BAR (0x%x)\n",
			    puc_port_type_name(sc->sc_desc->ports[i].type),
			    sc->sc_desc->ports[i].bar);
			continue;
		}

		/* set up to configure the child device */
		paa.port = i;
		paa.type = sc->sc_desc->ports[i].type;
		paa.flags = sc->sc_desc->ports[i].flags;
		paa.pc = pa->pa_pc;
		paa.tag = pa->pa_tag;
		paa.intrhandle = intrhandle;
		paa.a = sc->sc_bar_mappings[barindex].a +
		    sc->sc_desc->ports[i].offset;
		paa.t = sc->sc_bar_mappings[barindex].t;
		paa.dmat = pa->pa_dmat;
		paa.dmat64 = pa->pa_dmat64;

		if (!is_console &&
		    bus_space_subregion(sc->sc_bar_mappings[barindex].t,
		    sc->sc_bar_mappings[barindex].h,
		    sc->sc_desc->ports[i].offset,
		    sc->sc_bar_mappings[barindex].s -
		      sc->sc_desc->ports[i].offset,
		    &subregion_handle) != 0) {
			aprint_error_dev(self, "couldn't get subregion for port %d\n", i);
			continue;
		}
		paa.h = subregion_handle;

#if 0
		printf("%s: port %d: %s @ (index %d) 0x%x (0x%lx, 0x%lx)\n",
		    device_xname(self), paa.port,
		    puc_port_type_name(paa.type), barindex, (int)paa.a,
		    (long)paa.t, (long)paa.h);
#endif

		locs[PUCCF_PORT] = i;

		/* and configure it */
		sc->sc_ports[i].dev = config_found_sm_loc(self, "puc", locs,
			&paa, puc_print, config_stdsubmatch);
	}
}

CFATTACH_DECL_NEW(puc, sizeof(struct puc_softc),
    puc_match, puc_attach, NULL, NULL);

static int
puc_print(void *aux, const char *pnp)
{
	struct puc_attach_args *paa = aux;

	if (pnp)
		aprint_normal("%s at %s", puc_port_type_name(paa->type), pnp);
	aprint_normal(" port %d", paa->port);
	return (UNCONF);
}

const struct puc_device_description *
puc_find_description(pcireg_t vend, pcireg_t prod, pcireg_t svend,
    pcireg_t sprod)
{
	int i;

#define checkreg(val, index) \
    (((val) & puc_devices[i].rmask[(index)]) == puc_devices[i].rval[(index)])

	for (i = 0; puc_devices[i].name != NULL; i++) {
		if (checkreg(vend, PUC_REG_VEND) &&
		    checkreg(prod, PUC_REG_PROD) &&
		    checkreg(svend, PUC_REG_SVEND) &&
		    checkreg(sprod, PUC_REG_SPROD))
			return (&puc_devices[i]);
	}

#undef checkreg

	return (NULL);
}

static const char *
puc_port_type_name(int type)
{

	switch (type) {
	case PUC_PORT_TYPE_COM:
		return "com";
	case PUC_PORT_TYPE_LPT:
		return "lpt";
	default:
		panic("puc_port_type_name %d", type);
	}
}
